### PR TITLE
Adds pre-commit and linting support

### PR DESCRIPTION
**Pull request by CyclingNinja**

### CODE REVIEW

File: {{ cookiecutter.package_name }}/.pre-commit-config.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+exclude: "\\.asdf$"
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.pre-commit-config.yml:8]: Is this meant to have one repo?

File: {{ cookiecutter.package_name }}/.readthedocs.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+formats:
+  - htmlzip
+
+python:
+  install:
+    - method: pip
+      extra_requirements:
+        - all
+        - docs
+        - docs-gallery
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.readthedocs.yaml]: Did a separate gallery build get added to the template?

File: {{ cookiecutter.package_name }}/.rtd-environment.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+name: rtd_sunpy
+channels:
+  - conda-forge
+dependencies:
+  - python=3.10
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.rtd-environment.yml:5]:
```suggestion
  - python=3.12
```
Time to update this?

File: {{ cookiecutter.package_name }}/.rtd-environment.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  - pip
+  - openjpeg
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.rtd-environment.yml]:
```suggestion
```
I think this is only needed by core?

File: {{ cookiecutter.package_name }}/.rtd-environment.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+name: rtd_sunpy
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.rtd-environment.yml]: Should this be templated by the package name?

File: {{ cookiecutter.package_name }}/.pre-commit-config.yml (hunk 1 above)
Comment (CyclingNinja) [reply {{ cookiecutter.package_name }}/.pre-commit-config.yml:8]: :facepalm: it is

File: {{ cookiecutter.package_name }}/.rtd-environment.yml (hunk 5 above)
Comment (CyclingNinja) [reply {{ cookiecutter.package_name }}/.rtd-environment.yml]: yeah it should

File: {{ cookiecutter.package_name }}/.rtd-environment.yml (hunk 3 above)
Comment (CyclingNinja) [reply {{ cookiecutter.package_name }}/.rtd-environment.yml:5]: Given that the min python requirement is 3.10, is it not better to have the env run off that so we know it works for users?

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -1,5 +1,12 @@
 repos:
-  - repo: https://github.com/pre-commit/pre-commit-hooks
+  repos:
+    # This should be before any formatting hooks like isort
```
Comment (nabobalis) [inline .pre-commit-config.yaml]: Are you planning to add isort to the template or will we leave that until we adopt ruff more?

File: {{ cookiecutter.package_name }}/.ruff.toml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+# Implicit-namespace-package. The examples are not a package.
+"docs/*.py" = ["INP001"]
+"__init__.py" = ["E402", "F401", "F403"]
+"test_*.py" = ["B011", "D", "E402", "PGH001", "S101"]
+# Need to import clients to register them, but don't use them in file
+"sunpy/net/__init__.py" = ["F811"]
+# These files are allowed to use warnings.warn
+"sunpy/util/exceptions.py" = ["TID251"]
+"sunpy/util/tests/test_logger.py" = ["TID251"]
+"sunpy/util/decorators.py" = ["TID251"]
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.ruff.toml]:
```suggestion
```
I think these are only relevant to sunpy.

File: {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst [hunk 8]
```diff
@@ -1,10 +1,10 @@
-This project is based upon the Astropy package template
-(https://github.com/astropy/package-template/) which is licenced under the terms
+This project is based upon the Open Astronomy package template
+(https://github.com/open-astronomy/package-template/) which is licenced under the terms
 of the following licence.
 
 ---
 
-Copyright (c) 2018, Astropy Developers
+Copyright (c) 2018, Open Astronomy Developers
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst]: Wait, Open Astronomy isnt OpenAstronomy???!!!!

My entire life has been a lie.

File: {{ cookiecutter.package_name }}/.rtd-environment.yml (hunk 3 above)
Comment (nabobalis) [reply {{ cookiecutter.package_name }}/.rtd-environment.yml:5]: Yeah that's a good point.

File: .pre-commit-config.yaml (hunk 6 above)
Comment (CyclingNinja) [reply .pre-commit-config.yaml]: I am planning to add isort, I started putting this pre-commit here then doubted whether I should. It definitely belongs in the template, but is it needed here? Probably?

File: {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst (hunk 8 above)
Comment (CyclingNinja) [reply {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst]: No i'm just an eejit :joy:

File: .pre-commit-config.yaml (hunk 6 above)
Comment (Cadair) [reply .pre-commit-config.yaml]: I think we need to use isort not ruff as ruff doesn't support length sorting the stdlib and I don't want to generate the churn (especially on core) of changing that.

File: {{ cookiecutter.package_name }}/.readthedocs.yaml (hunk 2 above)
Comment (Cadair) [reply {{ cookiecutter.package_name }}/.readthedocs.yaml]: No, we haven't galleried yet

File: {{ cookiecutter.package_name }}/.flake8 [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    W504
+max-line-length = 110
+exclude =
+    .git
+    __pycache__
+    docs/conf.py
+    build
+    sunpy/data/sample.py
+    sunpy/extern/**
+    sunpy/__init__.py,
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.flake8]: core specific remove.

```suggestion
```

File: {{ cookiecutter.package_name }}/.ruff.toml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+target-version = "py310"
+line-length = 110
+exclude = [
+  ".git,",
+  "__pycache__",
+  "build",
+  "sunpy/extern/**",
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.ruff.toml]:
```suggestion
```

File: {{ cookiecutter.package_name }}/.ruff.toml [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  "sunpy/version.py",
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.ruff.toml]: needs templating based on package name and if dev versions are enabled.

File: {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst
Comment (Cadair) [inline {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst:1]: This should be done in a PR to upstream.

File: {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst [hunk 12]
```diff
@@ -1,10 +1,10 @@
-This project is based upon the Astropy package template
-(https://github.com/astropy/package-template/) which is licenced under the terms
+This project is based upon the Open Astronomy package template
+(https://github.com/open-astronomy/package-template/) which is licenced under the terms
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/licenses/TEMPLATE_LICENSE.rst]:
```suggestion
(https://github.com/OpenAstronomy/package-template/) which is licensed under the terms
```

**Review by Cadair — APPROVED** [review]

**Pull request merged**